rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py

```diff
@@ -0,0 +1,105 @@
+        """
+        if tx_log.topics[0] == SWAP_SIGNATURE:
+            # When the router chains multiple swaps in one transaction only the last swap has
+            # the buyer in the topic. In that case we know it is the last swap and the receiver is
+            # the user
+            maybe_buyer = hex_or_bytes_to_address(tx_log.topics[2])
+            out_event = in_event = None
+
+            # amount_in is the amount that enters the pool and amount_out the one
+            # that leaves the pool
+            amount_in_0, amount_in_1 = hex_or_bytes_to_int(tx_log.data[0:32]), hex_or_bytes_to_int(tx_log.data[32:64])  # noqa: E501
+            amount_out_0, amount_out_1 = hex_or_bytes_to_int(tx_log.data[64:96]), hex_or_bytes_to_int(tx_log.data[96:128])  # noqa: E501
+            amount_in, amount_out = amount_in_0, amount_out_0
+            if amount_in == ZERO:
+                amount_in = amount_in_1
+            if amount_out == ZERO:
+                amount_out = amount_out_1
+
+            received_eth = ZERO
+            for event in decoded_events:
+                if event.asset == A_ETH and event.event_type == HistoryEventType.RECEIVE:
+                    received_eth += event.balance.amount
+
+            for event in decoded_events:
+                # When we look for the spend event we have to take into consideration the case
+                # where not all the ETH is converted. The ETH that is not converted is returned
+                # in an internal transaction to the user.
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    (
+                        event.balance.amount == (spend_eth := asset_normalized_value(amount_in, event.asset)) or  # noqa: E501
+                        event.asset == A_ETH and spend_eth + received_eth == event.balance.amount
+                    )
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.SPEND
+                    event.counterparty = CPT_UNISWAP_V2
+                    event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in uniswap-v2 from {event.location_label}'  # noqa: E501
+                    out_event = event
+                elif (
+                    (maybe_buyer == transaction.from_address or event.asset == A_ETH) and
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.balance.amount == asset_normalized_value(amount_out, event.asset)
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.RECEIVE
+                    event.counterparty = CPT_UNISWAP_V2
+                    event.notes = f'Receive {event.balance.amount} {event.asset.symbol} in uniswap-v2 from {event.location_label}'  # noqa: E501
+                    in_event = event
+                elif (
```

Comment:
Didn't go deep in details, but question about the if-elifs here. What if `event_type` is `receive` and `amount` == `amount_out` but `(maybe_buyer == transaction.from_address or event.asset == A_ETH)` is false. Or it can't happen?

Reply:
If that happens then it's not a log we care about.

Reply:
I believe he means that we have to make sure that what we receive is ETH and not other thing (make a swap in favor of address Y and a send of LDO to that address Y in the same transaction)